src/foam/nanos/crunch/UserCapabilityJunctionDAO.js

```diff
@@ -105,13 +108,18 @@ foam.CLASS({
         UserCapabilityJunction old = (UserCapabilityJunction) super.find_(x, ucj.getId());
 
         // do not allow updates to sourceId/targetId properties
-        if ( old != null && ucj.getSourceId() != old.getSourceId() ) throw new RuntimeException("User's capability cannot be reassigned.");
-        if ( old != null && ! ucj.getTargetId().equals(old.getTargetId()) ) throw new RuntimeException("Capability cannot be changed.");
+        if ( old != null && ucj.getSourceId() != old.getSourceId() ) throw new RuntimeException(this.ERROR_TWO);
+        if ( old != null && ! ucj.getTargetId().equals(old.getTargetId()) ) throw new RuntimeException(this.ERROR_THREE);
         
         // if ucj data is set but does not match expected data, do not put
         Capability capability = (Capability) ucj.findTargetId(x);
-        if ( capability.getOf() != null && ucj.getData() != null && ! ( ucj.getData().getClassInfo().equals(capability.getOf()) ) )
-          throw new RuntimeException("Capability data provided is not of the correct type.");
+        if ( 
+          capability == null ||
+          (
+            ! ( capability.getOf() == null || ucj.getData() == null ) &&
+            ! ( ucj.getData().getClassInfo().equals(capability.getOf()) )
+          )
+        ) throw new RuntimeException(this.ERROR_FOUR + ucj.getTargetId());
```

Comment:
If this is something user can't do anything about, then 
1. create an Alarm - this will also generate an error log message.
2. throw the exception